codec/multicodecs.go

```diff
@@ -0,0 +1,84 @@
+//
+// This registry map is only used for default behaviors.
+// If you don't want to rely on it, you can always construct your own LinkSystem.
+// (For this reason, there's no special effort made to detect conflicting registrations in this map.
+// If more than one package registers for the same multicodec indicator, and
+// you somehow end up with both in your import tree, and yet care about which wins:
+// then just don't use this registry anymore: make a LinkSystem that does what you need.)
+var MulticodecEncoderRegistry = make(map[uint64]ipld.Encoder)
+
+// MulticodecDecoderRegistry is a simple map which maps a multicodec indicator number
+// to an ipld.Decoder function.
+//
+// Packages which implement an IPLD codec and have a multicodec number reserved in
+// https://github.com/multiformats/multicodec/blob/master/table.csv
+// are encouraged to register themselves in this map at package init time.
+// (Doing this at package init time ensures this map can be accessed without race conditions.)
+//
+// The linking/cid.DefaultLinkSystem will use this map to find decoders
+// to use when deserializing data from storage.
+//
+// This registry map is only used for default behaviors.
+// If you don't want to rely on it, you can always construct your own LinkSystem.
+// (For this reason, there's no special effort made to detect conflicting registrations in this map.
+// If more than one package registers for the same multicodec indicator, and
+// you somehow end up with both in your import tree, and yet care about which wins:
+// then just don't use this registry anymore: make a LinkSystem that does what you need.)
+var MulticodecDecoderRegistry = make(map[uint64]ipld.Decoder)
+
+// MultihashRegistry is a simple map which maps a multihash indicator number
+// to a standard golang Hash interface.
+//
+// Hashers which are available in the golang stdlib are registered here automatically.
+// Some hashes from x/crypto are also included out-of-the-box.
+//
+// Packages which want to register more hashing functions and have a multihash number reserved in
+// https://github.com/multiformats/multicodec/blob/master/table.csv
+// are encouraged to do so at package init time.
+// (Doing this at package init time ensures this map can be accessed without race conditions.)
+//
+// The linking/cid.DefaultLinkSystem will use this map to find decoders
+// to use when deserializing data from storage.
+//
+// This registry map is only used for default behaviors.
+// If you don't want to rely on it, you can always construct your own LinkSystem.
+// (For this reason, there's no special effort made to detect conflicting registrations in this map.
+// If more than one package registers for the same multicodec indicator, and
+// you somehow end up with both in your import tree, and yet care about which wins:
+// then just don't use this registry anymore: make a LinkSystem that does what you need.)
+var MultihashRegistry = make(map[uint64]func() hash.Hash)
```

Comment:
I would like to understand more the rationale for HasherChooser and this registry, along with the code in LinkPrototype.BuildLink

It feels like a reinvention of go-multihash. How do I know every codec already registered with go-multihash is registered here? What if they are missing and I don't expect that?

Why does BuildLink take a hash sum instead of the bytes of the whole node? Instead of writing to a hash function returned by HasherChooser, why not just write to a byte buffer and then give the bytes to buildLink which can then just call cid.Prefix.Sum in the cidLink implementation?

I feel like this pushes us further into the "reinvent low level libraries" direction. While I'm open enough to the possibility of a go-cid rewrite to allow for a Link abstraction (though I'm unconvinced still about the trade-off in refactor effort, even with the link interface abstraction), I'm definitely opposed to trying to rewrite go-multihash, which has only a few dependencies. In any case, if you have Link.BuildLink take all the bytes rather than a hash, you remove all these hash library depedencies any time you want to import the codec table for example.

Reply:
It is.  go-multihash is frankly awful.

I covered some of this in the initial PR message, but with greater :fire: :

go-multihash makes it very difficult to handle multihash identifiers in a forward compatible way (impossible, in fact), which is ridiculous.  It has direct unconditional transitive dependencies which many projects don't need, which adds friction to adopting IPLD as a whole, and needs to be fixed.  It doesn't work streamingly, which is just ridiculous.  It doesn't provide a way to use a bog-standard `hash.Hash` without pushing patches (and a surprising number of them, because the core abstractions are terrible) directly into that library, which is also ridiculous.  It doesn't even connect with go-cid particularly well (it predates a refactor which uses strings for immutability and map-key reasons), which is (ready?) ridiculous.  There's very little that's good about the current state of go-multihash.

I did not completely remove go-multihash from the dependency tree in this diffset yet, but it's close, and I probably should, either now or soon.  One of the recommendations I got from Jeromy when looking at the BuildLink method is that we could just cut out the middleman go-mulithash code, assemble the bytes ourselves, and use one of the go-cid parse methods.  It's a somewhat staggering idea, but would probably honestly be more efficient than some of the dancing go-multihash does right now.

I don't doubt that this migration could be bumpy, because it *does* mean less things are registered by default.  But I'm highly confident that getting a registry in this area somewhere is the way forward, and has to happen sometime (JS has already gone the direction of reworking these primitives so dependencies are optional, and we should too); and I think this is probably going to be one of the easier times to do it.

I'll add some more packaging to this to make the migration easier -- I'm thinking of making a `go-ipld-prime/multihash/all` package which, when imported, makes sure everything currently registered by go-multihash is registered here too.  Any application that's a data broker that needs to support all the hashes can import that and be done.  Hopefully that and a good changelog message will be enough.

Reply:
(I'll confess, I didn't start out with this much aggression towards go-multicodec.  It developed during the writing of this PR, when I got hands-on with it.  I did *not* expect it to be so hard to take an interface draft using golang's standard-since-the-first-release-of-golang `hash.Hash` and connect it with doing hashing using a library with the word hash in the name.  And it.  was.  awful.  Hair-pulling, screaming-at-the-screen awful.  Simple things should be simple.  This should have been, and was not.)

It's possible that we should recuse up to renovating the go-multihash repo.  However, what I came to in the course of working on this PR seems to address most of the painpoints in a way that's low impact to downstreams, other than requiring a new import statement.  And there's a lot of reasons I'm happy with it:

- I don't think the upsides of dynamic registration are realizable with anything less.
- I actually got this PR _done_.  I don't think changing go-multicodec is on the same timescale.
  - And I don't really want to couple this change to that one, because I want us to be able to start writing new code against LinkSystem *now*.
- If we do push these semantics back towards go-multihash later: the effect to downstreams of go-ipld-prime can be as mild as changing their import statement from `go-ipld-prime/multihash/all` to `go-multihash/register/all`.
  - Or less: the `go-ipld-prime/multihash/all` package can just itself be reduced to an import that passes it on, leaving downstream users will no breaking changes at all.
  
So: are there smells of workaround here: yes.  But I think they're minimal and manageable; and the workarounds that would've been generated in this PR continuing to use go-multihash directly would actually be larger and less manageable.